nexus/temporal-payloads.md

```diff
@@ -0,0 +1,74 @@
+# Temporal Payloads over Nexus+HTTP
+
+## Motivation
+
+For ideal integration of Temporal applications with Nexus, Temporal users should be able to leverage Temporal Payloads
+and not have to concern themselves with Nexus implementation details.
+
+When a workflow completes, we'd like Temporal server to be able to deliver the completion result to any registered Nexus
+callbacks and the callback receiver (a Temporal server) to translate the completion to a Temporal Payload.
+
+Inherently, we need a fixed set of translation rules from Payloads to Nexus+HTTP and back. These rules should be
+executed server-side.
+
+## Background
+
+Nexus reserves the entire HTTP request and response bodies (for input and successful results) for data and uses HTTP
+headers (a mapping of string to an array of strings) for transmitting metadata.
+
+> The main motivation for this design was to support streaming and avoid requiring HTTP proxies to inspect request
+> bodies.
+
+A Temporal Payload is defined as:
+
+```proto3
+message Payload {
+    map<string,bytes> metadata = 1;
+    bytes data = 2;
+}
+```
+
+## Translation rules
+
+To unmarshal a Temporal Payload from a Nexus request or response:
+
+1. Extract all `Content-Temporal-` prefixed headers into the payload metadata using the following rules:
+   1. Ignore multiple values (not supported)
+   2. Strip the key prefix and transform header case to camel case. E.g: `Content-Temporal-Foo-Bar` becomes `fooBar`
+   3. Read the values as [data URLs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs)
+2. Set the `encoding` Payload metadata field as follows:
+   1. If `Content-Length` is 0 or body is empty → `binary/null`
+   2. If `Content-Type` media type is `application/json` →
+      1. If the content type parameter contains a `format=protobuf` → `json/protobuf` and set `messageType` to the corresponding parameter value.
+         Example: `Content-Type: application/json; format=protobuf; messageType=com.example.Message`
+      2. Otherwise → `json/plain`
+   3. If `Content-Type` media type is `application/x-protobuf` → `binary/protobuf` and set the `messageType` parameter as given in the header.
+      Example: `Content-Type: application/x-protobuf; messageType="com.example.Message"`
+   4. Otherwise → `binary/plain`
+3. Extract all `Content-Encoding` values into a `codecs` metadata key with a comma separator. Note that the server may
```

Comment:
I think this is premature and doesn't need to exist today (nor does `Accept-Encoding`). I think we can decide who converts HTTP content later from these headers (IMO it would be the Nexus side not the Temporal side reading these new metadata keys).

Reply:
Agree this could be deferred but I'm already thinking ahead to provide some information to users that want to know which codecs to apply.
This will also apply to encryption in my mind, e.g: you may see the following sequence `gzip, aes` and you'd need to deflate and decrypt.

Reply:
I don't think we should confuse application codecs w/ HTTP ones. Let the HTTP middleware/proxies handle known HTTP codecs, and let applications handle known application codecs.

Reply:
Fair point.

But I do think that we'll want to come up with some convention for "application" codecs considering that these applications are going to be handling requests by arbitrary callers.

Reply:
I think you'll want to affect the content type. Not only would one expect to not leak the content type of their encoded body (so it should be embedded in the encoded body), but content type can't really be JSON if it's not JSON. This can sound like an argument for reusing `Content-Encoding` but I still think that deserves to be separate and that still leaks the internal content type.

Yes, it can be annoying deciding how to embed the original content type in the body, but that's the encoder authors problem since they have to extract it back out. Can also discuss whether other metadata deserves to be able to be encoded-and-hidden.

Reply:
I don't think they want to embed the original content type in the body, they can use headers for that.
Other metadata binary and encoded as base64, it's up to the the custom data converter implementation to put that metadata there in the first place.

Reply:
Agreed offline to have the HTTP server handle content encoding.